### PR TITLE
dashboard and applet changes

### PR DESCRIPTION
**Pull request by fanmingyu212**
Partially addresses #48. This PR depends on https://github.com/Jayich-Lab/pydux/pull/1942.

List of changes:
* Adds modified versions of the experiment explorer and the scheduler (originally in artiq) to jax.tools.applets. These applets are designed to work with the artiq and the parameter bank servers.
  - By default, experiments now use parameters when they are scheduled (preload parameters), rather than when they are prepared (previous behavior). This can be overridden by right-click on the experiment and go to details. The preloaded parameters are temporarily saved in the parameter bank before the experiment queries the parameters. If they are stopped through the experiment schedule GUI, the temporarily saved parameter cache is cleared too.
  - Adds a function to scan the device database in the experiment explorer. This should be used with caution. If the device names are changed, you may need to run the initialization experiment again and restart the dds applet.
  - Allows running the initialization and the background experiments by right-clicking in the experiment schedule GUI.
  - Adds a `parameters_updated` signal in the explorer, so the parameter bank *can* use this signal to show only relevant parameters (also see comments later).
* Modifies the parameter loading function in `JaxEnvironment` to support preloading parameters.
* Adds `jax.tools.dashboard.control_board` and `jax.tools.dashboard.plot_board`. The only difference between them is how the applets are organized in the board. The control board centers a small blank area, which allows us to place applets around it like before. The ARTIQ logo was removed from the blank area. The plot board does not have the blank area for maximizing the space to display the plots. See attached screenshots.
* Modifies the DDS applet for visual changes. It should work better for high DPI windows now.
* PMT applet code style change.

This PR closes the first issue in #48 (experiment uses the parameters when scheduled), and makes addressing most of the other issues in #48 easier. However, this PR does not fully allow addressing synchronizing the experiment selected in the explorer and the parameters in the parameter bank. We will need the two applets to directly talk to each other (initializing them in the control board) for that. However, this is blocked by how we implement labrad with applets, as we cannot run multiple reactors in a single program while these applets want to make separate labrad connections. This can be solved by implementing labrad differently, but it is probably better to skip this issue for now.

### CODE REVIEW

**Comment by fanmingyu212** [comment]
Plot board:
plot_board

**Comment by fanmingyu212** [comment]
Control board:
control_board

File: base/environments/jax_environment.py [hunk 1]
```diff
@@ -284,17 +284,23 @@ def save_parameters(self):
...
         from jax.util.labrad import remove_labrad_units
-
-        params = {}
-        params_full = {}
         pb = self.cxn.parameter_bank
+        params_serialized, params_full_serialized = pb.get_frozen_parameters(self.scheduler.rid)
+        params = pickle.loads(params_serialized)
+        params_full = pickle.loads(params_full_serialized)
         for collection, name in self.parameter_paths:
             if collection not in params:
```
Comment (mczimmermann) [inline base/environments/jax_environment.py:292]: if collection is not in params, should it throw an error?

File: base/environments/jax_environment.py [hunk 2]
```diff
@@ -284,17 +284,23 @@ def save_parameters(self):
         Also saves all parameters to the data file.
         """
         from jax.util.labrad import remove_labrad_units
-
-        params = {}
-        params_full = {}
         pb = self.cxn.parameter_bank
+        params_serialized, params_full_serialized = pb.get_frozen_parameters(self.scheduler.rid)
+        params = pickle.loads(params_serialized)
+        params_full = pickle.loads(params_full_serialized)
```
Comment (mczimmermann) [inline base/environments/jax_environment.py:290]: change this name? It's unclear what params_full is (maybe cached_params or copied_params?)

File: tools/applets/dds.py [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
     # widgets can only be created in the default thread.
     do_initialize = QtCore.pyqtSignal()
 
     def __init__(self, args, **kwds):
         super().__init__(**kwds)
-        self.setDisabled(True)  # start with the applet disabled, until artiq server is connected.
+        self.setObjectName("DDS")
+        self.setDisabled(
+            True
+        )  # start with the applet disabled, until artiq server is connected.
```
Comment (mczimmermann) [inline tools/applets/dds.py:22]: this is never set back to false within the file?

Comment (fanmingyu212) [reply tools/applets/dds.py:22]: In line 58 it is set back to `False`.

File: base/environments/jax_environment.py (hunk 1 above)
Comment (fanmingyu212) [reply base/environments/jax_environment.py:296]: I added some docstrings to clarify.

File: base/environments/jax_environment.py (hunk 2 above)
Comment (fanmingyu212) [reply base/environments/jax_environment.py:294]: Also added docstrings to clarify.

File: tools/applets/dds_channel.py [hunk 4]
```diff
@@ -164,7 +163,7 @@ def initialize_gui(self):
         self.freq_box.setSingleStep(0.1)
         self.freq_box.setFont(spinboxfont)
         self.freq_box.setKeyboardTracking(False)
-        MHz_to_Hz = 1.e6
+        MHz_to_Hz = 1.0e6
```
Comment (mczimmermann) [inline tools/applets/dds_channel.py:166]: not sure what we usually do for this but would it be easier to define it as a global or class constant? It seems we use it 3+ times

File: tools/applets/schedule.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        self.cxn.add_on_connect("artiq", self.run_in_labrad_loop(self.artiq_connected))
+        self.cxn.add_on_disconnect("artiq", self.artiq_disconnected)
+
+    async def artiq_connected(self):
+        self.artiq = self.cxn.get_server("artiq")
+        if not self._disconnect_reported:
+            self.setDisabled(False)
+
+    def artiq_disconnected(self):
+        self.setDisabled(True)
```
Comment (mczimmermann) [inline tools/applets/schedule.py:127]: should this also call self._report_disconnect()

File: tools/applets/schedule.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    async def request_term_multiple(self, rids):
+        """Terminates multiple experiments."""
+        for rid in rids:
+            try:
+                await self.artiq.request_terminate_experiment(rid)
+            except Exception as e:
+                # May happen if the experiment has terminated by itself
+                # while we were terminating others.
+                logger.warning(
+                    "failed to request termination of RID %d", rid, exc_info=True
```
Comment (mczimmermann) [inline tools/applets/schedule.py:164]: it may be helpful to add what you have in the comment in the actual error message, ex: "failed to request termination of RID %d, experiment may have terminated itself while in queue"

File: tools/dashboard/control_board.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        qfm = QtGui.QFontMetrics(self.font())
+        self.resize(140 * qfm.averageCharWidth(), 38 * qfm.lineSpacing())
+
+        self.exit_request = asyncio.Event()
+
+    def closeEvent(self, event):
+        event.ignore()
+        self.exit_request.set()
+
+    def save_state(self):
```
Comment (mczimmermann) [inline tools/dashboard/control_board.py:82]: should we add a button that does this? I'm a little confused about the current use for it

**Comment by mczimmermann** [comment]
this is out of the scope of this PR but should I open an issue to update the tray launcher to the new dashboard?

File: tools/applets/schedule.py (hunk 6 above)
Comment (fanmingyu212) [reply tools/applets/schedule.py:164]: The code is mostly copied from artiq (as said in the docstring of the class). The message here and comments are directly from artiq too. I prefer not to add extra information as there can be multiple reasons that an experiment cannot be terminated (e.g. artiq server is off, artiq master is down, Sinara devices are disconnected).

File: tools/dashboard/control_board.py (hunk 7 above)
Comment (fanmingyu212) [reply tools/dashboard/control_board.py:82]: This function (and restoreState) is called when the dashboard initializes. This is a part of PyQt for restoring and saving the state of the window. Again the code is copied from artiq

File: tools/applets/schedule.py (hunk 5 above)
Comment (fanmingyu212) [reply tools/applets/schedule.py:127]: I don't think this function should call report_disconnected. The function is called when the artiq server disconnects. report_disconnected is for when the artiq subscribers disconnect.

File: tools/applets/dds_channel.py (hunk 4 above)
Comment (fanmingyu212) [reply tools/applets/dds_channel.py:166]: I agree. This is out of the scope of this PR though.

**Comment by fanmingyu212** [comment]
@mczimmermann we do not need an issue to update the tray launcher scripts to run the new dashboards. We can just modify the tray launcher scripts on each computer.

**Comment by mczimmermann** [comment]
sounds good! this looks good

**Review by mczimmermann — APPROVED** [review]
